File: ios/Classes/LibrsyncNativePlugin.swift
import Flutter
import UIKit

public class LibrsyncNativePlugin: NSObject, FlutterPlugin {
  var eventChannelMap = [String: FlutterEventChannel]()
  var eventSinkMap = [String: FlutterEventSink]()
  static var FlutterPluginRegistrar registrar?;

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "librsync_native/call", binaryMessenger: registrar.messenger())
    let instance = LibrsyncNativePlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  // Define the Stream protocol
  protocol Stream {
    func send(_ bytes: [UInt8], _ length: Int)
  }

  // Define a Swift class that conforms to the Stream protocol
  class CallbackStream: Stream {
    let callback: ([Uint8]) -> Void

    init(callback: @escaping ([Uint8] -> Void)) {
      self.callback = callback
    }

    func send(_ bytes: [UInt8], _ length: Int) {
        callback(bytes)
    }
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "startStream":
      let arguments = call.arguments as? [String: Any]
      let streamID = arguments["streamID"] as? String
      let eventChannel = FlutterEventChannel(name: streamID, binaryMessenger: LibrsyncNativePlugin.registrar!.messenger())
      eventChannelMap[streamID] = eventChannel
      eventChannel.setStreamHandler(EventHandler(streamID: streamID), callback: { streamID, eventSink in
        eventSinkMap[streamID] = eventSink
      })
    case "computeDelta":
      let arguments = call.arguments as? [String: Any]
      let signatureStr = arguments["signatureStr"] as? String
      let targetPath = arguments["targetPath"] as? String
      let streamID = arguments["streamID"] as? String

      let callbackStream = CallbackStream(callback: { _ bytes: [Uint8] in 
        let result: [String: Any]()
        result["bytes"] = bytes
        let eventSink = eventSinkMap[streamID]
        eventSink.success(result)
      })


    default:
      result(FlutterMethodNotImplemented)
    }
  }

  
}

class EventHandler : FlutterStreamHandler {
  let streamID: String
  let callback: (String, FlutterEventSink) -> Void
  init(streamID: String, callback: @escaping (String, FlutterEventSink) -> Void) {
    self.streamID = streamID
    self.callback = callback
  }

  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    callback(self.streamID, eventSink)
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    return nil
  }
}

File: android/src/main/java/me/gurupras/librsyncnative/librsync_native/LibrsyncNativePlugin.java
package me.gurupras.librsyncnative.librsync_native;

import android.util.Log;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

import androidx.annotation.NonNull;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import librsyncbridge.Stream;

/** LibrsyncNativePlugin */
public class LibrsyncNativePlugin implements FlutterPlugin, MethodCallHandler {
  static final String TAG = "librsync-native-java";
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private MethodChannel channel;
  private FlutterPluginBinding flutterPluginBinding;
  private Map<String, EventChannel.EventSink> eventSinkMap;
  private Map<String, EventChannel> eventChannelMap;

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    GoLogger.init();
    this.flutterPluginBinding = flutterPluginBinding;
    eventSinkMap = new HashMap<>();
    eventChannelMap = new HashMap<>();
    channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "librsync_native/call");
    channel.setMethodCallHandler(this);
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    if (call.method.equals("startStream")) {
      Log.d(TAG, "Received startStream request");
      final String streamID = call.argument("streamID");
      EventChannel eventChannel = new EventChannel(flutterPluginBinding.getBinaryMessenger(), streamID);
      eventChannel.setStreamHandler(new EventChannel.StreamHandler() {
        @Override
        public void onListen(Object arguments, EventChannel.EventSink events) {
          eventSinkMap.put(streamID, events);
          Log.d(TAG, "Received listener for EventChannel: " + streamID);
        }

        @Override
        public void onCancel(Object arguments) {
          eventSinkMap.remove(streamID);
        }
      });
      eventChannelMap.put(streamID, eventChannel);
      result.success(null);
    } else if (call.method.equals("computeDelta")) {
      String signatureStr = call.argument("signatureStr");
      String targetPath = call.argument("targetPath");
      final String streamID = call.argument("streamID");
      
      EventChannel eventChannel = eventChannelMap.get(streamID);
      librsyncbridge.Stream stream = new Stream() {
        @Override
        public void send(byte[] bytes, long l) {
          EventChannel.EventSink sink = eventSinkMap.get(streamID);
          Map<String, Object> eventData = new HashMap<>();
          eventData.put("bytes", bytes);
          Log.d(TAG, "Received bytes to send to eventChannel. streamID=" + streamID + " bytes=" + l);
          if (sink != null) {
            sink.success(eventData);
          }
        }
      };
      librsyncbridge.Librsyncbridge.computeDelta(signatureStr, targetPath, stream);
      // TODO: Maybe close the streams?
      eventChannelMap.remove(streamID);
      eventSinkMap.remove(streamID);
      result.success(null);
    }
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }
}
